gridappsd/simulation.py

```diff
@@ -1,63 +1,140 @@
-        if 'logMessage' in message:
-            log_message = message['logMessage']
-            if log_message.startswith("done with timestep"):
-                timestep = int(log_message[len("done with timestep "):])
-                if self._timestep_finished_callback is not None:
-                    self._timestep_finished_callback(self, timestep)
-                # We use -1 here because the last timestep from fncs-goss-bridge.py never actually
-                # comes through at least as of 2019-6-7
-                if timestep >= self._duration-1:
-                    self._is_running = False
-
-        if self._show_timesteps:
-            # print(headers, message)
-            # print(headers)
-            sys.stdout.write(f"{message['logMessage']}\n")
-            # print(message["logMessage"])
-
-    @property
-    def running(self):
-        return self.is_running()
-
-    @property
-    def showing_status(self):
-        return self._show_timesteps
+    """ Simulation object allows controlling simulations through a python API.
+
+    The simulation object allows controlling and monitoring of simulations through
+    a python API.  It is capable of starting, stopping, pausing and restarting simulations
+    that are run on GridAPPSD.
+
+    There are three events that can be monitored by multiple callbacks:
+
+        - timestep:
+
+            This event is triggered each time a timestep has been triggered
+
+        - on_timestep_begin:
+
+            This event is triggered when a new tiemtep has been started.
+
+        - on_finished:
+
+            This event is triggered when the simulation has completed.
+
+    """
+    def __init__(self, gapps, run_config):
+        # Protect from circular import.
+        from . import GridAPPSD
+        assert isinstance(gapps, GridAPPSD), "Must be an instance of GridAPPSD"
+        assert isinstance(run_config, dict)
+        self._gapps = gapps
+        self._run_config = deepcopy(run_config)
+
+        # Will be populated when the simulation is first started.
+        self.simulation_id = None
+
+        self.__on_start = set()
+        self.__on_next_timestep_callbacks = set()
+        self.__on_simulation_complete_callbacks = set()
+
+        self._measurement_count = 0
+        self._log_count = 0
+        self._platform_log_count = 0
+
+        self._num_timesteps = run_config['simulation_config']['duration']
+
+        # Devices that the user wants measurements from
+        self._device_measurement_filter = {}
+
+        self.__filterable_measurement_callback_set = set()
+
+    def start_simulation(self):
+        """ Start the configured simulation by calling the REQUEST_SIMULATION endpoint.
+        """
+        resp = self._gapps.get_response(t.REQUEST_SIMULATION, json.dumps(self._run_config))
+
+        if 'simulationId' not in resp:
+            message = "Simulation was not able to run\n" + str(resp)
+            raise Exception(message)
```

Comment:
@craig8 - it would probably be best practice here to use an explicit exception rather than the generic and high-level "Exception."

Something like:
```python
class SimulationFailedToStartError(Exception):
    """Exception raised if a simulation fails to start."""
    pass
```

and then:
```python
raise SimulationFailedToStartError(message)
```

This certainly isn't critical, but from my understanding would be more in line with Python best practices.